src/daft-parquet/src/metadata.rs

```diff
@@ -12,11 +16,220 @@ fn metadata_len(buffer: &[u8], len: usize) -> i32 {
+                let field_id = field_info.id;
+                if let Some(mapped_field) =
+                    field_id.and_then(|field_id| self.field_id_mapping.get(&field_id))
+                {
+                    field_info.name = mapped_field.name.clone();
+                }
+
+                let fields = match logical_type {
+                    // GroupTypes with logical types List/Map have intermediate child nested fields without field_ids,
+                    // but we need to keep recursing on them to keep renaming on their children.
+                    Some(_) => fields,
+                    // GroupTypes without logical_types are just structs, and we can go ahead with removing
+                    // any fields that don't have field IDs with a corresponding match in the mapping
+                    None => fields
+                        .into_iter()
+                        .filter(|f| {
+                            f.get_field_info()
+                                .id
+                                .and_then(|field_id| self.field_id_mapping.get(&field_id))
+                                .is_some()
+                        })
+                        .collect(),
+                };
+
+                Ok(ParquetTypeWrapper(ParquetType::GroupType {
+                    field_info,
+                    logical_type,
+                    converted_type,
+                    fields,
+                }))
+            }
+        }
+    }
+}
+
+/// Applies field_ids to a ParquetType, returns `None` if the field ID is not found
+/// in the `field_id_mapping`, or if the field has no field ID.
+fn apply_field_ids_to_parquet_type(
+    parquet_type: ParquetType,
+    field_id_mapping: &BTreeMap<i32, Field>,
+) -> Option<ParquetType> {
+    let field_id = parquet_type.get_field_info().id;
+    if field_id
+        .and_then(|field_id| field_id_mapping.get(&field_id))
+        .is_some()
+    {
+        let mut pq_type_rewriter = ParquetTypeFieldIdRewriter { field_id_mapping };
+        let rewritten_pq_type = ParquetTypeWrapper(parquet_type)
+            .rewrite(&mut pq_type_rewriter)
+            .unwrap()
+            .0;
+        Some(rewritten_pq_type)
+    } else {
+        None
+    }
+}
+
+/// Applies field_ids to a parquet2 FileMetaData struct
+/// 1. Rename columns based on the `field_id_mapping`
+/// 2. Drop any columns that don't have a `field_id`, or that don't have a corresponding entry in the `field_id_mapping`
+fn apply_field_ids_to_parquet_file_metadata(
+    file_metadata: FileMetaData,
+    field_id_mapping: &BTreeMap<i32, Field>,
+) -> super::Result<FileMetaData> {
+    use parquet2::metadata::{ColumnChunkMetaData, SchemaDescriptor};
+
+    // Construct a new Parquet Schema after applying renaming using field_ids
+    let new_schema_descr = SchemaDescriptor::new(
+        file_metadata.schema_descr.name().to_string(),
+        file_metadata
+            .schema_descr
+            .fields()
+            .iter()
```

Comment:
you can take ownership with `into_iter` and would the clone in `apply_field_ids_to_parquet_type`

Reply:
or use https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each

Reply:
Unfortunately the SchemaDescriptor doesn't give ownership of its private inner `.fields`, so I can't take ownership of it (only iterate on `.fields() -> &[ParquetType]`)